tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs

```diff
@@ -0,0 +1,370 @@
+                logger.LogInformation($"MSAL Token source is : { result.AuthenticationResultMetadata.TokenSource }");
+                return result.AccessToken;
+            }
+            catch (MsalException ex)
+            {
+                logger.LogError(ex.Message);
+                return ex.Message;
+            }
+        }
+
+        /// <summary>
+        /// Clear default request headers on the http client
+        /// </summary>
+        /// <param name="logger"></param>
+        /// <returns></returns>
+        private static void ClearDefaultRequestHeaders(ILogger logger)
+        {
+            logger.LogInformation("ClearDefaultRequestHeaders Function called.");
+
+            s_httpClient.DefaultRequestHeaders.Clear();
+        }
+
+        /// <summary>
+        /// Sets the authorization header on the http client
+        /// </summary>
+        /// <param name="token"></param>
+        /// <param name="logger"></param>
+        /// <returns></returns>
+        private static void SetAuthorizationHeader(
+            string? token,
+            ILogger logger)
+        {
+            logger.LogInformation("SetAuthorizationHeader Function called.");
+
+            s_httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
+        }
+
+        /// <summary>
+        /// Gets a JSON String using Newtonsoft JSON Convert
+        /// </summary>
+        /// <param name="content"></param>
+        /// <returns></returns>
+        public static string ConvertToJsonString(object? content)
+        {
+            string json;
+
+            json = JsonConvert.SerializeObject
+                (content, Formatting.Indented, new JsonSerializerSettings()
+                { ReferenceLoopHandling = ReferenceLoopHandling.Serialize, TypeNameHandling = TypeNameHandling.None });
+
+            return json.Trim();
+
+        }
+
+        /// <summary>
+        /// Returns Content Result for final output from the web api
+        /// </summary>
+        /// <param name="content"></param>
+        /// <param name="contentEncoding"></param>
+        /// <param name="statusCode"></param>
+        /// <returns></returns>
+        public static ContentResult GetContentResult(string content, string contentEncoding, int statusCode)
+        {
+            if (statusCode == (int)MSIHelper.HTTPErrorResponseCode.Status200OK)
+            { 
+                ManagedIdentityResponse? managedIdentityResponse = JsonConvert.DeserializeObject<ManagedIdentityResponse>(content);
+
+                //return ManagedIdentityResponse object so we can trim the access token
+                return new ContentResult { 
+                    Content = ConvertToJsonString(managedIdentityResponse), 
+                    ContentType = contentEncoding, 
+                    StatusCode = statusCode 
+                };
+            }
+
+            //return errors as is from the MSI Endpoints
+            return new ContentResult { 
```

Comment:
What about Http exceptions?

Reply:
this should return all error messages and exceptions from MSI 

**Sample validation message from MSI :** 

`curl -X 'GET' \
  'https://service.msidlab.com/GetMSIToken?uri=http%3A%2F%2F127.0.0.1%3A41292%2Fmsi%2Ftoken%2F&azureResource=webapp' \
  -H 'accept: text/plain' \
  -H 'X-IDENTITY-HEADER: 69C62B109AAF4EB7B01061197C14F550'`

`"{\"error\":{\"code\":\"ApiVersionUnspecified\",\"message\":\"An API version is required, but was not specified.\",\"innerError\":null}}"`

**Wrong API Version :** 

`curl -X 'GET' \
  'https://service.msidlab.com/GetMSIToken?uri=http%3A%2F%2F127.0.0.1%3A41292%2Fmsi%2Ftoken%2F%3Fapi-version%3D2019-01-01%26resource%3Dhttps%3A%2F%2Fmanagement.azure.com&azureResource=webapp' \
  -H 'accept: text/plain' \
  -H 'X-IDENTITY-HEADER: 69C62B109AAF4EB7B01061197C14F550'`

`"{\"error\":{\"code\":\"UnsupportedApiVersion\",\"message\":\"The HTTP resource that matches the request URI 'http://127.0.0.1:41292/msi/token' does not support the API version '2019-01-01'.\",\"innerError\":null}}"`

**Exceptions**

`curl -X 'GET' \
  'https://service.msidlab.com/GetMSIToken?uri=http%3A%2F%2F127.0.0.1%3A41292%2Fmsi%2Ftoken%2F%3Fapi-version%3D2019-08-01%26resource%3Dhttps%3A%2F%2Fmanagement.azure.com&azureResource=function' \
  -H 'accept: text/plain' \
  -H 'X-IDENTITY-HEADER: 69C62B109AAFEB7B01061197C14F550'`

`"An attempt was made to access a socket in a way forbidden by its access permissions. (127.0.0.1:41292) \n\nAn attempt was made to access a socket in a way forbidden by its access permissions."
`

Reply:
With the error codes

![image](https://user-images.githubusercontent.com/90415114/206876897-f4074075-583d-4ef5-b86d-20717004f06d.png)